[Discuss] Avoid IEnumeration allocation (List+Dictionary)

benaadams opened the pull request:
(For discussion) 

1M enumerations from 1M objects at 40 MBytes to 5 objects at 1,224 bytes
Current impl takes 846ms and has 12 Gen0 collections (x3.25 slower than struct)
This impl takes 714ms and has 0 GC collections +15% faster than current (x2.7 slower than struct)
(Struct enumerator takes 260ms and has 0 GC collections)

The common `IEnumerator` interface `foreach` code

``` csharp
IList<int> list = new List<int>();
foreach (var value in list)
{
    // ...
}
```

Turns into something like

``` csharp
IEnumerable<int> enum = list.GetEnumerator();
try
{
    while (enum.MoveNext())
    {
        var value = enum.Current;
        // ...
    }
}
finally
{
    enum.Dispose();
}
```

Which boxes the struct enumerator to the interface and causes allocation.

This change introduces `ObjectPool<T>`; closely based on `ArrayPool<T>` - as an `internal` type (so as not to expose new api).

`List<T>` and `Dictionary<TKey, TValue>` use this shared pool for class enumerators which contain the struct enumerator. This allows no boxing/allocation (beyond the initial pool allocations); when the interface is used for List and Dictionary enumeration. 

The enumerator is returned to the pool on the call to Dispose and the struct enumerator set to `default(Enumerator)`.

Interface dispatch will still be slower than using the stuct enumerator however.

Resolves dotnet/roslyn#4446 (Measurements below)

/cc @rynowak, @stephentoub, @davidfowl, @jaredpar, @JonHanna

benaadams committed:
Avoid IEnumeration allocation (List+Dictionary)

benaadams commented:
Allocations for https://gist.github.com/benaadams/d05000ae4b4935aa3e3b39ce47713a6d
```csharp
var list = new List<int>();
var ilist = (IList<int>)list;

for (var i = 0; i < 100; i++)
{
    list.Add(i);
}

for (var i = 0; i < 1000000; i++)
{
    foreach(var value in ilist)
    {
        ;
    }
}
```

Before 1M objects at 40 MBytes

![before allocs](http://aoa.blob.core.windows.net/aspnet/enumerator1.png)

After 5 objects at 1,224 bytes

![after allocs](http://aoa.blob.core.windows.net/aspnet/enumerator2.png)

mikedn commented on src/mscorlib/src/System/Collections/Generic/DefaultObjectPool.cs:
```diff
@@ -0,0 +1,102 @@
+            _objects = new T[maxPooled];
+        }
+
+        /// <summary>Takes an object from the pool. If the pool is empty, returns a new object.</summary>
```
How is the new object created?

benaadams replied:
Its just `new T()`; I'm investigating something more generic with a lamda

khellang replied:
Look at [line 68](https://github.com/dotnet/coreclr/pull/4468/files#diff-f9ba5645900085107330a8012452edc7R68); `obj ?? new T();`

mikedn replied:
Yeah, missed the `?? new T()`. Costly due to the use of `Activator.CreateInstance`.

mikedn commented on src/mscorlib/src/System/Collections/Generic/DefaultObjectPool.cs:
```diff
@@ -0,0 +1,102 @@
+            // While holding the lock, grab whatever is at the next available index and
+            // update the index.  We do as little work as possible while holding the spin
+            // lock to minimize contention with other threads.  The try/finally is
+            // necessary to properly handle thread aborts on platforms which have them.
```
This code isn't thread abort safe if you care about that. `index` may be incremented without setting the array element to null. Then in `Return` `index` can be decremented without setting the array element to the returned object. Next call to `Rent` will return an object that's still in use.

stephentoub replied:
Just for reference, this comment was copied from ArrayPool in corefx.  The concern there wasn't in making the pool in general thread abort safe, but rather just ensuring that the spin lock wouldn't be abandoned in the presence of a thread abort; the concern was about causing other threads to then livelock indefinitely rather than in ensuring that all state remained consistent.

mikedn commented on src/mscorlib/src/System/Collections/Generic/List.cs:
```diff
@@ -1051,6 +1054,29 @@ public void Sort(IComparer<T> comparer)
+            object IEnumerator.Current => Current;
+            public void Dispose()
+            {
+                _enumerator = default(Enumerator);
```
This break existing code that accidentally uses an enumerator after `Dispose` has been called. In the existing implementation both `MoveNext` and `Current` continue to work after the enumerator is disposed. In the new implementation `MoveNext` throws `NullReferenceException` and `Current` returns `default(T)`.

benaadams replied:
Is that common? You'd have to go via the interface rather than the concrete type and manually be calling Dispose. It would be hard to do the pooling otherwise without changing the compiler and doing escape analysis.

mikedn replied:
I don't think it's common but it certainly is a behavior change that needs to be considered.

davidroth commented on src/mscorlib/src/System/Collections/Generic/DefaultObjectPool.cs:
```diff
@@ -0,0 +1,102 @@
+                if (lockTaken) _lock.Exit(false);
+            }
+
+            return obj ?? new T();
```
`new T()` will use `System.Activator` to create new instances. So i guess this will slow down things if the pool hast not yet been populated or if all items are in use and new ones are created. (Although I must admit that this is speculative, I have not measured this). Anyway, maybe provide a cached  `Func<T>` to the `Rent`method, so that it can create new objects without falling back to activator?

```csharp
public override T Rent(Func<T> activator)
{
  ....
return obj ?? activator();
}

// List<T> Code:
IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator() 
{
    var enumerator = ObjectPool<PooledIEnumerator>.Shared.Rent(() => new PooledIEnumerator()); // delegate should be cached in a static field.
    return enumerator.Initalize(this);
}
```

mattwarren commented:
As the potential cost of `new T()` has come up a few times, here's a [BenchmarkDotNet](https://github.com/PerfDotNet/BenchmarkDotNet) benchmark ([full benchmark code](https://gist.github.com/mattwarren/3a41f741d1900f73545a49022212116f)):

```ini
BenchmarkDotNet=v0.9.4.0
OS=Microsoft Windows NT 6.1.7601 Service Pack 1
Processor=Intel(R) Core(TM) i7-4800MQ CPU @ 2.70GHz, ProcessorCount=8
Frequency=2630751 ticks, Resolution=380.1196 ns, Timer=TSC
HostCLR=MS.NET 4.0.30319.42000, Arch=32-bit RELEASE
JitModules=clrjit-v4.6.100.0

Type=Program  Mode=Throughput  Runtime=Clr LaunchCount=1  
```
                  Method |     Median |    StdDev | Scaled |
------------------------ |----------- |---------- |------- |
 ComplexClassAlternative | 86.8452 ns | 6.4257 ns |  13.17 |
     ComplexClassRegular |  8.8911 ns | 0.0960 ns |   1.35 |
  SimpleClassAlternative | 82.9110 ns | 0.5299 ns |  12.57 |
      SimpleClassRegular |  6.5960 ns | 0.3710 ns |   1.00 |

Note: Regular is `new SomeClass()` and Alternative is `new T()`

So there is a cost, but it's roughly 80 **nanoseconds** (I'm not sure whether that's a problem or not?)

benaadams commented:
@mattwarren and taking a parameter `Func<T>` which is `() => new SomeClass()`?

mattwarren commented:
@benaadams Not sure I'm following, could you update the gist with the code you want to test and I'll run it for you?

mattwarren commented:
Here's the updated results, `Func<T>` which is `() => new SomeClass()` is much quicker:

```ini

BenchmarkDotNet=v0.9.4.0
OS=Microsoft Windows NT 6.1.7601 Service Pack 1
Processor=Intel(R) Core(TM) i7-4800MQ CPU @ 2.70GHz, ProcessorCount=8
Frequency=2630751 ticks, Resolution=380.1196 ns, Timer=TSC
HostCLR=MS.NET 4.0.30319.42000, Arch=32-bit RELEASE
JitModules=clrjit-v4.6.100.0

Type=Program  Mode=Throughput  Runtime=Clr  
LaunchCount=1  

```
                  Method |     Median |    StdDev | Scaled |
------------------------ |----------- |---------- |------- |
 ComplexClassAlternative | 84.5888 ns | 0.6949 ns |  13.12 |
        ComplexClassFunc | 20.4752 ns | 0.0950 ns |   3.18 |
     ComplexClassRegular |  8.8755 ns | 0.1359 ns |   1.38 |
  SimpleClassAlternative | 82.2324 ns | 0.3058 ns |  12.75 |
         SimpleClassFunc | 18.3899 ns | 0.1852 ns |   2.85 |
      SimpleClassRegular |  6.4488 ns | 0.0486 ns |   1.00 |

benaadams commented:
Changed the pool to take a `Func<T>` and `Func<TState, T>, TState` in rent.
Changed `PooledIEnumerator` to be a generic type rather than defining a type in `List` and `Dictionary` so its only the two methods in each that change. (Though they do end up having a horrible number of angle brackets)

Measuring perf now

mikedn commented on src/mscorlib/src/System/Collections/Generic/List.cs:
```diff
@@ -567,16 +567,19 @@ bool System.Collections.IList.Contains(Object item)
-            return enumerator.Initalize(this);
+            var enumerator = ObjectPool<PooledIEnumerator<T, Enumerator, List<T>>>.Shared.Rent(
+                                () => new PooledIEnumerator<T, Enumerator, List<T>>());
+            return enumerator.Initalize((list) => new Enumerator(list), this);
```
This seems overly complicated. `enumerator._enumerator = new Enumerator(list)` would do the job.

benaadams replied:
yeah, good point!

Janiels commented on src/mscorlib/src/System/Collections/Generic/PooledIEnumerator.cs:
```diff
@@ -0,0 +1,34 @@
+        }
+        public void Reset()
+        {
+            throw new NotSupportedException();
```
Should this not call into `_enumerator.Reset()`? Right now it looks breaking (on a bigger level than just the return type change) since the list and dictionary enumerators support Reset.

benaadams replied:
Relying on reset via the interface is dangerous though as https://msdn.microsoft.com/en-us/library/system.collections.ienumerator.reset(v=vs.110).aspx

> The Reset method is provided for COM interoperability. It does not necessarily need to be implemented; instead, the implementer can simply throw a NotSupportedException.

So any collection passed using the interface may throw `NotSupportedException`; the struct/concrete enumerator's behaviour remains the same.

stephentoub replied:
That doesn't change the fact that lots of code could easily be relying on it; even if it comes via the interface, code could know that the only instances passed in to their code supported reset, and/or could be catching the exception to handle cases where it didn't.

benaadams replied:
I could always not use a pooled enumerator for the non-generic IEnumerator, which implements the Reset method (is hidden on the generic implementation)?

Janiels replied:
Taking `List<T>` as an example, all the enumerators returned are `List<T>.Enumerator` structures - just boxed for the explicit interface implementations. So for both interface implementation (`IEnumerable<T>` and `IEnumerable`) the enumerator returned supports `Reset`. However as you mention in the issue, someone could hold on to the enumerator after `Dispose` and reset it, which would cause issues in other places using the returned pooled allocator. I don't think that problem is solvable this way unfortunately.

benaadams replied:
Returned Reset

Janiels commented on src/mscorlib/src/System/Collections/Generic/DefaultObjectPool.cs:
```diff
@@ -0,0 +1,113 @@
+        }
+
+        /// <summary>Takes an object from the pool. If the pool is empty, returns a new object by calling the factory.</summary>
+        public override T Rent(Func<T> factory)
```
What about completely avoiding `factory` with a `public bool TryRent(out T value)` instead?

benaadams replied:
Added

khellang replied:
Except the `Func<T>` overloads are still there and the method isn't in use :wink:

benaadams replied:
`<sigh />` :smile: ok...

benaadams replied:
Sorted

benaadams replied:
Made it faster also :)

benaadams commented:
The current implementation takes 780ms and has 12 Gen0 collections
This implementation take 715ms and has 0 GC collections

benaadams commented:
Main issue remaining so far, is no one trusts people to stop using objects post-`Dispose`; and a finalizer would be too slow.

benaadams commented:
Maybe if the Enumerators were Destructible Types? https://github.com/dotnet/roslyn/issues/161

benaadams commented:
Maybe if the Enumerators were Destructible Types? https://github.com/dotnet/roslyn/issues/161
Don't know if that would work with re-usable destruction though...

benaadams commented:
or if you could `move this` https://github.com/dotnet/roslyn/issues/160
```csharp
public void Dispose()
{
    Enumerator = default(TEnumerator);
    ObjectPool<PooledIEnumerator<TValue, TEnumerator>>.Shared.Return(move this);
}
```

benaadams commented:
Closing this as seen better solutions (e.g. @jaredpar's enumerator)
